### DISCUSSION TITLE
Create form with specific columns

### DISCUSSION THREAD

**Discussion by markplus** [nasirkhan/laravel-starter#469] [Q&A] [answered]
BUT if I want to create a form with a table set up with specific fields, and not with the default ones how can I do? I'm trying I created the form with the command you described and then I modified the table but now I don't see anything in the administrator side table list, in the public side I see the inserted fields but there is nothing in the administrative side table, I removed the name, slug, description field and I put the columns I need but they are not visible.
index_datatable.blade.php

```
@extends('backend.layouts.app')

@section('title') {{ __($module_action) }} {{ __($module_title) }} @endsection

@section('breadcrumbs')
<x-backend-breadcrumbs>
    <x-backend-breadcrumb-item type="active" icon='{{ $module_icon }}'>{{ __($module_title) }}</x-backend-breadcrumb-item>
</x-backend-breadcrumbs>
@endsection

@section('content')
    <div class="card">
        <div class="card-body">

            <x-backend.section-header>
                <i class="{{ $module_icon }}"></i> {{ __($module_title) }} <small class="text-muted">{{ __($module_action) }}</small>

                <x-slot name="subtitle">
                    @lang(":module_name Management Dashboard", ['module_name'=>Str::title($module_name)])
                </x-slot>
                <x-slot name="toolbar">
                    @can('add_'.$module_name)
                    <x-buttons.create route='{{ route("backend.$module_name.create") }}' title="{{__('Create')}} {{ ucwords(Str::singular($module_name)) }}" />
                    @endcan

                    @can('restore_'.$module_name)
                    <div class="btn-group">
                        <button class="btn btn-secondary dropdown-toggle" type="button" data-coreui-toggle="dropdown" aria-expanded="false">
                            <i class="fas fa-cog"></i>
                        </button>
                        <ul class="dropdown-menu">
                            <li>
                                <a class="dropdown-item" href='{{ route("backend.$module_name.trashed") }}'>
                                    <i class="fas fa-eye-slash"></i> @lang("View trash")
                                </a>
                            </li>
                            <!-- <li>
                                <hr class="dropdown-divider">
                            </li> -->
                        </ul>
                    </div>
                    @endcan
                </x-slot>
            </x-backend.section-header>

            <div class="row mt-4">
                <div class="col">
                    <table id="datatable" class="table table-bordered table-hover table-responsive-sm">
                        <thead>
                            <tr>
                                <th>
                                    #
                                </th>
                                <th>
                                    @lang("ticket::text.identificativo")
                                </th>
                                <th>
                                    @lang("ticket::text.updated_at")
                                </th>
                                <th class="text-end">
                                    @lang("ticket::text.action")
                                </th>
                            </tr>
                        </thead>
                    </table>
                </div>
            </div>
        </div>
        <div class="card-footer">
            <div class="row">
                <div class="col-7">
                    <div class="float-left">

                    </div>
                </div>
                <div class="col-5">
                    <div class="float-end">

                    </div>
                </div>
            </div>
        </div>
    </div>

@endsection

@push ('after-styles')
    <!-- DataTables Core and Extensions -->
    <link rel="stylesheet" href="{{ asset('vendor/datatable/datatables.min.css') }}">

@endpush

@push ('after-scripts')
    <!-- DataTables Core and Extensions -->
    <script type="module" src="{{ asset('vendor/datatable/datatables.min.js') }}"></script>

    <script type="module">
        $('#datatable').DataTable({
            processing: true,
            serverSide: true,
            autoWidth: true,
            responsive: true,
            ajax: '{{ route("backend.$module_name.index_data") }}',
            columns: [{
                    data: 'id',
                    name: 'id'
                },
                {
                    data: 'identificativo',
                    name: 'identificativo'
                },
                {
                    data: 'updated_at',
                    name: 'updated_at'
                },
                {
                    data: 'action',
                    name: 'action',
                    orderable: false,
                    searchable: false
                }
            ]
        });
    </script>
@endpush
```

**Comment by nasirkhan** [comment]
If you have checked the Backend controller of the new module, there are no methods. If you want to add/update columns in the index page, you have to add the index_datatable method in the backend controller as well. 

Check the BackendBaseController and you will get the methods.

**Reply by markplus** [reply]
I did everything, modified index_datatable.blade, index.blade, form.blade, if I try to create a new one it is created and written correctly in the database, fronted side I see everything correctly, but backend side I don't see anything in the table, the table is completely empty no fields are displayed

**Comment by nasirkhan** [comment, marked as answer]
You have to add the following method  in your controller

```php
public function index_data()
    {
        $module_title = $this->module_title;
        $module_name = $this->module_name;
        $module_path = $this->module_path;
        $module_icon = $this->module_icon;
        $module_model = $this->module_model;
        $module_name_singular = Str::singular($module_name);

        $module_action = 'List';

        $page_heading = label_case($module_title);
        $title = $page_heading.' '.label_case($module_action);

        $$module_name = $module_model::select('id', 'name', 'updated_at');

        $data = $$module_name;

        return Datatables::of($$module_name)
            ->addColumn('action', function ($data) {
                $module_name = $this->module_name;

                return view('backend.includes.action_column', compact('module_name', 'data'));
            })
            ->editColumn('name', '<strong>{{$name}}</strong>')
            ->editColumn('updated_at', function ($data) {
                $module_name = $this->module_name;

                $diff = Carbon::now()->diffInHours($data->updated_at);

                if ($diff < 25) {
                    return $data->updated_at->diffForHumans();
                } else {
                    return $data->updated_at->isoFormat('llll');
                }
            })
            ->rawColumns(['name', 'action'])
            ->orderColumns(['id'], '-:column $1')
            ->make(true);
    }
```

then you have to update the following line and include the columns you want to display in the index table

```php
$$module_name = $module_model::select('id', 'name', 'updated_at');
```

then you have to update the `index_datatable.blade.php` file and update the following js block as per your column names. 

```js
<script type="module">
    $('#datatable').DataTable({
        processing: true,
        serverSide: true,
        autoWidth: true,
        responsive: true,
        ajax: '{{ route("backend.$module_name.index_data") }}',
        columns: [{
                data: 'id',
                name: 'id'
            },
            {
                data: 'name',
                name: 'name'
            },
            {
                data: 'updated_at',
                name: 'updated_at'
            },
            {
                data: 'action',
                name: 'action',
                orderable: false,
                searchable: false
            }
        ]
    });
</script>
```

**Comment by markplus** [comment]
> You have to add the following method in your controller
> 
> ```
> public function index_data()
>     {
>         $module_title = $this->module_title;
>         $module_name = $this->module_name;
>         $module_path = $this->module_path;
>         $module_icon = $this->module_icon;
>         $module_model = $this->module_model;
>         $module_name_singular = Str::singular($module_name);
> 
>         $module_action = 'List';
> 
>         $page_heading = label_case($module_title);
>         $title = $page_heading.' '.label_case($module_action);
> 
>         $$module_name = $module_model::select('id', 'name', 'updated_at');
> 
>         $data = $$module_name;
> 
>         return Datatables::of($$module_name)
>             ->addColumn('action', function ($data) {
>                 $module_name = $this->module_name;
> 
>                 return view('backend.includes.action_column', compact('module_name', 'data'));
>             })
>             ->editColumn('name', '<strong>{{$name}}</strong>')
>             ->editColumn('updated_at', function ($data) {
>                 $module_name = $this->module_name;
> 
>                 $diff = Carbon::now()->diffInHours($data->updated_at);
> 
>                 if ($diff < 25) {
>                     return $data->updated_at->diffForHumans();
>                 } else {
>                     return $data->updated_at->isoFormat('llll');
>                 }
>             })
>             ->rawColumns(['name', 'action'])
>             ->orderColumns(['id'], '-:column $1')
>             ->make(true);
>     }
> ```
> 
> then you have to update the following line and include the columns you want to display in the index table
> 
> ```
> $$module_name = $module_model::select('id', 'name', 'updated_at');
> ```
> 
> then you have to update the `index_datatable.blade.php` file and update the following js block as per your column names.
> 
> ```js
> <script type="module">
>     $('#datatable').DataTable({
>         processing: true,
>         serverSide: true,
>         autoWidth: true,
>         responsive: true,
>         ajax: '{{ route("backend.$module_name.index_data") }}',
>         columns: [{
>                 data: 'id',
>                 name: 'id'
>             },
>             {
>                 data: 'name',
>                 name: 'name'
>             },
>             {
>                 data: 'updated_at',
>                 name: 'updated_at'
>             },
>             {
>                 data: 'action',
>                 name: 'action',
>                 orderable: false,
>                 searchable: false
>             }
>         ]
>     });
> </script>
> ```
thanks, great
